### PR TITLE
Adding Linux compatibility, improved and up to date software detection as well as Unreal Engine 5 support

### PR DESCRIPTION
**Pull request by gpijat**
Hi @huntfx,

I've sent you a message on LinkedIn about this, I hope you're doing well.

For this pull request I would like to share a new way to detect the current software, it keeps the base idea you were using, based on if yes or no the software module can be imported and a test on the sys.executable string.
Instead of testing for a substring match, it's now based on regular expression.
I wrote basic pattern for each software and made some more advanced one for the one I were able to do more tests. At the end, I don't think there's much need for such complexe regex for the maya validation but I had them and I though it could just be even more accurate.
The use of regex allows 2 things
- Linux and possible MacOs support (Actually, I can't test on MacOs, so this is theoretical)
- Unreal Engine 5 support and backward compatibility with Unreal Engine 4

This Pull request also change a few things for substance painter, where docking were not working as expected, add a `QtCore.Qt.Dialog` flag for the houdini window as it seems to be the only way to keep it in foreground, above the main window, without forcing it always on top which would make it annoying.
I've also moved the custom exception to a new module so we should have more granularity and we can use in different modules, which I do.

I hope this will work for you, feel free to give me any comment or retake and I would be happy to discuss. 

PS: Also, if we go ahead with this, could it be possible to update the module on pypi ? When using pip, we get a really old version that doesn't work.

### CODE REVIEW

**Comment by huntfx** [comment]
Hey, thanks a bunch for the PR :)

I'll start testing this out at work, though in the meantime I'll do a pypi release on the current version, hadn't realised it was that far behind. Also speaking of pip installs, add a version check on your end to make sure it'll stay on `1.*`, I'll likely do some much needed changes with 2.0 as I've been living with my early bad design choices for years haha.

Here's some things that have come to mind from scanning down the changes:

- The `find_spec` / `find_loader` import checks (rather than `try/except import`) _may_ have been because one program was a bit funny about importing their own modules on startup. Don't remember if that was the case but I'll check everything still works.
- Was `closeEvent` meant to be removed from Substance Painter? The idea is it'll save the window position on close with non docked windows so it can restore the position. I'll test this too.
- Could I request renames for `vfxExceptions` and `vfxSoftware` to `vfxwindow/exceptions.py` and `vfxwindow/utils/software.py`? Just personal preference here but I feel like the `vfx` prefix is a bit redunant, and the user being able to more easily import the exceptions module (as in not have it hidden in `utils`) could come in useful.
- Love the regex idea. The downside of your more "complete" Maya's Windows regex is that it won't support custom install locations - while it's likely to be installed in Program Files it's not guaranteed (not sure if the same applies to Linux's `/usr`?)
- (Just putting this here to remind myself to properly test your Maya icon rewrite)

Anyway thanks again so much, we're still mostly using Maya/Nuke at work so I haven't had chance to give the other applications much love or testing, so this is definitely appreciated.

**Comment by gpijat** [comment]
Hey !

Sure thing, I'll update the version to the next minor, my bad 👍 

- The `find_spec` / `find_loader` change I did was for 2 reasons: Actually, they seem to import the module to know if it's importable, and then, they access the `__spec__` attribute. Maya, Houdini, Substance Painter & Designer seems to be ok with that, but for unreal, the value in this attribute is `None` and this error is raised:
```py
LogPython: Error: Traceback (most recent call last):
LogPython: Error:   File "<string>", line 1, in <module>
LogPython: Error:   File "/opt/EpicGames/Linux_Unreal_Engine_5.3.0/Engine/Binaries/ThirdParty/Python3/Linux/lib/python3.9/importlib/util.py", line 114, in find_spec
LogPython: Error:     raise ValueError('{}.__spec__ is None'.format(name))
LogPython: Error: ValueError: unreal.__spec__ is None
```
So because they import the module anyway I went "why not try to import it myself, should be easier". But I would be curious if you can find which soft have trouble with this method :)

- I removed the closeEvent because it was duplicated, there was 2 implementation for the same method in the `SubstancePainterWindow`. I though the other one seems to be the "good" one. But I may be wrong. Feel free to correct me.

- Sure thing, I'll rename those module and do as you suggest. The main reason for prefixing them is to reduce risk for name clash when importing other modules, like the python standard `exceptions` module. Anyway, there's still namespace to prevent this, I'll do this change :)

- Actually, you're right, the complex regexes won't support custom install, they are built around maya's default installation, and I tested them at my current job, without surprise, they don't work. That's why there's multiple regexes and that they go from very specific to more permissive, the code will fallback on each one of them one by one. For custom install the first one won't work but the one checking for `maya\.(?:bin|exe|app)` should work fine whether it's the default install or a custom install.

Thanks you so much for your feedback, I'll update this soon with your requests :)

I would love if you can test on Nuke, and if you can send me what `sys.executable` return I would love it, it should allow me to improve the regex. Right now I was only able to test:
- Maya (Linux / Windows)
- Unreal Engine (Windows / Linux -> It doesn't work on linux, actually, I wasn't able to make PySide or PyQt work at all, still need test. But I may also be the only one trying to use UE5 on Linux.. ^^' )
- Houdini (Linux / Windows)
- Substance Painter (Linux)
- Substance Designer (Linux)
- Blender (Windows)

I'll run some more test anyway, I need to test in standalone for both Linux and Windows.
Also, I'll try to update the documentation to specify where it has been tested. For instance I'm using Linux Mint, based on Ubuntu/Debian, so I can't really tell what it looks like on other distros. I'll try on Arch soon if I can make maya works 🤞

**Comment by huntfx** [comment]
Thanks again, will get this running on my PC at work tomorrow. My Linux knowledge is next to none by the way, but if each distro is effectively a new OS, perhaps something like this could work to show the tests?

|  | Windows | Linux Mint | Ubuntu |
| ---------------------- | ------- | ---------- | ------ |
| Maya                  | :heavy_check_mark: | :grey_question: | :grey_question: |
| Nuke                  | :heavy_check_mark: | :x: | :x: |
| Blender               | :heavy_check_mark: | :grey_question: | :grey_question: |
| Substance Painter     | :heavy_check_mark: | :grey_question: | :grey_question: |
| Substance Designer    | :heavy_check_mark: | :grey_question: | :grey_question: |
| Houdini               | :heavy_check_mark: |  :heavy_check_mark: | :x: |
| Unreal Engine 4              | :heavy_check_mark: | :grey_question: | :grey_question: |
| Unreal Engine 5              | :heavy_check_mark: | :grey_question: | :grey_question: |

With the name clashes, they shouldn't ever be an issue anyway since everything is within the `vfxwindow` namespace and uses relative imports. My bad about the duplicate `closeEvent` method too, I have a vague suspicion I noticed it before but was busy working on something else and then completely forgot about it lol.

Shame about Unreal not working, I've got UE5 at work so can give it a test on Windows at least. If you'd like to try debug by completely bypassing `vfxwindow`, I believe it's more or less doing this behind the scenes (skipping `app.exec_()` as it probably locked up the interface):
```
try:
    app = QtWidgets.QApplication(sys.argv)
except RuntimeError:
    app = QtWidgets.QApplication.instance()
mainWindow().show()
```

Here's the Nuke executables.
```
C:\\Program Files\\Nuke9.0v9\\Nuke9.0.exe
C:\\Program Files\\Nuke12.1v3\\Nuke12.1.exe
C:\\Program Files\\Nuke14.0v5\\Nuke14.0.exe
```

**Comment by gpijat** [comment]
Actually, most Linux version should work the same, especially as they share a lot. But there are always some dark distribution out there that will want to do everything differently x) 
I think we can say `Linux*` as it should work on most distributions, especially those used in VFX, like Red Hat, CentOS, Fedora. Or the one users usually use like Ubuntu, Mint, maybe Arch as well.
So having `Linux*` Would allow us to mention below the tab `* Tested on Linux Mint.`.

`I have a vague suspicion I noticed it before but was busy working on something else and then completely forgot about it lol` Ahaha, that's something that can happen, let me know if I kept the wrong one !

For UE5, I tried this, unfortunately, it doesn't work on Linux, only Windows :(
I can't figure out why tho, I tried a few different things, of course, app.exec_() works but leave UE5 unresponsive.. So I tried running a QApplication in a different thread and it seems to work, both my UI and Unreal are responsive. But if I close I can't make it work again. And I don't think it will be easy to make it communicate with UE on the main thread either.

**Comment by gpijat** [comment]
Hey, just so you don't miss it, I addressed your request about the modules name and added a table in the README.md.

EDIT: Also updated the Nuke regex to make sure it works.

**Comment by huntfx** [comment]
Thanks, just giving it a few quick tests now then will leave it running.

Here are a few bugs I had to fix, I'll attempt to commit to the PR if it lets me:
- `NameError: global name 'ModuleNotFoundError' is not defined` - Sadly gotta keep that Python 2 compatibility with catching `ImportError` only
- `SubstancePainterWindow` is not imported
- The use of `vfxExceptions` wasn't renamed in `__init__.py`

One actual problem is it now breaks Nuke when running on startup again, but this [happened before ](https://github.com/huntfx/vfxwindow/commit/22347c2810831d165a4d3ef4352d66f8ee524754) so leave it with me.

All is working on Maya, Substance P+D and Blender, I don't have a Houdini license but that _should_ be fine, and when I get a bit of time I'll have to figure out UE5 😅 

And yes it appears you got the correct `closeEvent` lol, I think I was improving the docked window support and wrote a new method not realising there was an existing one.

**Comment by huntfx** [comment]
Ah yep UE5 (5.1.1) is working too with `Tools > Execute Python Script`.

Here's some code to do a quick test with by the way:
```
from vfxwindow.debug import TestWindow
TestWindow.show()
```

**Comment by gpijat** [comment]
All those thing that doesn't work are my mistake, I shouldn't have sent an update this late in the night. Those are just typos or missing "test" lines.
They are all fixed in the latest commit, except for the nuke startup thing of course :)

**Comment by huntfx** [comment]
Fixed the Nuke startup now, it was just a typo with `isNuke` making it always fail, causing `StandaloneWindow` to be used (which blocks the main GUI as it uses `app.exec_()`)

Anyway thanks so much bunch for the help, I'll leave this on my own PC and ensure no issues crop up, then I'll get it released to the studio as the big test.

**Comment by gpijat** [comment]
@huntfx That's a good news, how do we proceed now ? Should I pull those changes into my branch, can I merge my branch (Or should you do It ?).
That's my first time contributing to an open source project so I'm not aware how the process works.

Thank you ! :)

**Comment by huntfx** [comment]
To be honest this is the first PR I've received haha, this is quite new to me too. I was able to commit as if this was a branch, so I presume you should be able to pull down my changes?

I've merged this into the main branch now to bring the code up to version 1.8, but I'll let it run at work for a week or two just to check no issues crop up before doing the pypi release.

**Comment by gpijat** [comment]
@huntfx Awesome, I was happy to help, you're right to make more in depth tests before releasing it on pypi, I'll keep an eye on the repository to make sure I install the update :)

Let me know if there's anything, I will continue to use this project on some projects so if I have anything I will let you know as well through Issues and PR if needed.

**Comment by huntfx** [comment]
Yeah thanks, for the most part the "standalone" windows are stable so you should be all good on that front, the tricky part is when docking is used as every implementation works differently.

-----

Just as a heads up, not that I'm expecting you to lol, but if you _did_ want to try add support for any of the other software that I don't have access to, it's mostly as easy as copying and pasting an existing "simple" window, such as [Unreal](https://github.com/huntfx/vfxwindow/blob/master/vfxwindow/unreal.py) / [Fusion](https://github.com/huntfx/vfxwindow/blob/master/vfxwindow/fusion.py).

Basically just replace `VERSION` and `getMainWindow` if possible, ctrl+F to replace the software name, then in `show()`, modify `kwargs['instance']` and `kwargs['exec_']` to find a combination that works without locking things up. 

Again, not expecting you to that, just writing it down in case you regularly use another app and want to use this with it.
